File: neet_code_all_cpp/sliding_window/min_size_subarray_sum.cpp
#include <climits>
#include <iostream>
#include <vector>

class Solution {
    public:
        // AC
        // O(NlogN)
        int minSubArrayLen(int target, std::vector<int>& nums) {
            int n = nums.size();
            std::vector<int> prefix_sum(n + 1);
            prefix_sum[0] = 0;
            for(int i = 0; i < n; ++i) {
                prefix_sum[i + 1] = nums[i] + prefix_sum[i];
            }

            int result = INT_MAX;

            for(int i = 0; i <= n; ++i) {
                auto diff = target + prefix_sum[i];

                auto [left, right] = std::pair(0, n);
                while(left <= right) {
                    auto mid = left + (right - left) / 2;
                    if(prefix_sum[mid] == diff) {
                        result = std::min(result, std::abs(mid - i));
                        break;
                    }

                    if(prefix_sum[mid] < diff) {
                        left = mid + 1;
                    } else {
                        right = mid - 1;
                        result = std::min(result, std::abs(mid - i));
                    }
                }
            }

            if(result == INT_MAX) {
                return 0;
            }

            return result;
        }
};

// 模範解答
class SolutionAns {
    public:
        // AC
        int minSubArrayLen(int target, std::vector<int>& nums) {
            int fp = 0, sp = 1;
            int sum = nums[0];
            int min = INT_MAX;

            while(fp != sp) {
                if(sum >= target) {
                    min = std::min(sp - fp, min);
                    sum = sum - nums[fp];
                    ++fp;
                } else {
                    if(sp < nums.size()) {
                        sum += nums[sp];
                        ++sp;
                    } else {
                        ++fp;
                    }
                }
            }

            if(min == INT_MAX) {
                return 0;
            }

            return min;
        }
};

int main(void) {
    using Case = std::pair<int, std::vector<int>>;
    Case case_1 = {7, {2, 3, 1, 2, 4, 3}};
    // => 2
    Case case_2 = {4, {1, 4, 4}};
    // => 1
    Case case_3 = {11, {1, 1, 1, 1, 1, 1, 1, 1}};
    // => 0
    Case case_4 = {11, {1, 2, 3, 4, 5}};
    // => 3


    Solution s_1;
    std::cout << s_1.minSubArrayLen(case_1.first, case_1.second) << std::endl;
    std::cout << s_1.minSubArrayLen(case_2.first, case_2.second) << std::endl;
    std::cout << s_1.minSubArrayLen(case_3.first, case_3.second) << std::endl;
    std::cout << s_1.minSubArrayLen(case_4.first, case_4.second) << std::endl;

    SolutionAns s_ans;
    std::cout << s_ans.minSubArrayLen(case_1.first, case_1.second) << std::endl;
    std::cout << s_ans.minSubArrayLen(case_2.first, case_2.second) << std::endl;
    std::cout << s_ans.minSubArrayLen(case_3.first, case_3.second) << std::endl;
    std::cout << s_ans.minSubArrayLen(case_4.first, case_4.second) << std::endl;
}

File: neet_code_all/src/sliding_window/min_size_subarray_sum.rs
struct Solution {}
impl Solution {
    // O(NlogN)
    // AC
    pub fn min_sub_array_len(target: i32, nums: Vec<i32>) -> i32 {
        let n = nums.len() as i32;

        let mut perfix_sum = vec![0; n as usize + 1];
        perfix_sum[0] = 0;
        for i in 0..n {
            perfix_sum[i as usize + 1] = nums[i as usize] + perfix_sum[i as usize];
        }
        //println!("prefix_sum: {:?}", perfix_sum);
        let mut result = i32::MAX;
        for i in 0..=n {
            let diff = target + perfix_sum[i as usize];

            let (mut left, mut right) = (0, n as i32);
            while left <= right {
                let mid = left + (right - left) / 2;
                if perfix_sum[mid as usize] == diff {
                    result = result.min((mid - i).abs());
                    break;
                }

                if perfix_sum[mid as usize] < diff {
                    left = mid + 1;
                } else {
                    right = mid - 1;
                    result = result.min((mid - i).abs());
                }
            }
        }

        if result == i32::MAX {
            0
        } else {
            result
        }
    }
}

// AC
// C++の模範解答より
struct SolutionAnsCpp {}
impl SolutionAnsCpp {
    pub fn min_sub_array_len(target: i32, nums: Vec<i32>) -> i32 {
        let (mut fp, mut sp) = (0, 1);
        let mut sum = nums[0];
        let mut min = i32::MAX;

        while fp != sp {
            if sum >= target {
                min = min.min(sp - fp);
                sum -= nums[fp as usize];
                fp += 1;
            } else {
                if sp < nums.len() as i32 {
                    sum += nums[sp as usize];
                    sp += 1;
                } else {
                    fp += 1;
                }
            }
        }

        if min == i32::MAX {
            return 0;
        }

        min
    }
}

fn main() {
    let case_1 = (7, vec![2, 3, 1, 2, 4, 3]);
    // => 2
    let case_2 = (4, vec![1, 4, 4]);
    // => 1
    let case_3 = (11, vec![1, 1, 1, 1, 1, 1, 1, 1]);
    // => 0
    let case_4 = (11, vec![1, 2, 3, 4, 5]);
    // => 3

    println!(
        "case_1: {}",
        Solution::min_sub_array_len(case_1.0, case_1.1.clone())
    );
    println!(
        "case_2: {}",
        Solution::min_sub_array_len(case_2.0, case_2.1.clone())
    );
    println!(
        "case_3: {}",
        Solution::min_sub_array_len(case_3.0, case_3.1.clone())
    );
    println!(
        "case_4: {}",
        Solution::min_sub_array_len(case_4.0, case_4.1.clone())
    );

    println!(
        "case_1: {}",
        SolutionAnsCpp::min_sub_array_len(case_1.0, case_1.1.clone())
    );
    println!(
        "case_2: {}",
        SolutionAnsCpp::min_sub_array_len(case_2.0, case_2.1.clone())
    );
    println!(
        "case_3: {}",
        SolutionAnsCpp::min_sub_array_len(case_3.0, case_3.1.clone())
    );
    println!(
        "case_4: {}",
        SolutionAnsCpp::min_sub_array_len(case_4.0, case_4.1.clone())
    );
}
